sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -72,6 +72,11 @@ object Cast {
 
     case (StringType, CalendarIntervalType) => true
 
+    case (LongType, DayTimeIntervalType) => true
+    case (DayTimeIntervalType, LongType) => true
+    case (IntegerType, YearMonthIntervalType) => true
+    case (YearMonthIntervalType, IntegerType) => true
```

Comment:
I think we should support casting from/to any numeric types including Decimal for instance. And handle overflow properly.

Reply:
Do other databases support it? We don't encourage people to cast from numeric to datetime anymore, and we have created dedicated functions to do this.

cc @gengliangwang

Reply:
I would prefer to look at the SQL standard, first of all. It allows (requires) such casting as far as I know.

Reply:
I encounter this when I want to write test case in last pr about DayTimeIntervalType.
Since it store value as Long, so I think it can convert a long value to DayTimeIntervalType. But it failed.

Reply:
![image](https://user-images.githubusercontent.com/1097932/112946820-d24de780-90ea-11eb-8992-41075e7c1a51.png)
![image](https://user-images.githubusercontent.com/1097932/112946962-06c1a380-90eb-11eb-8deb-8c0d940ba66b.png)

Reply:
I just lookup the ANSI standard and it forbids converting intervals as Approximate Numeric types. How about let's support integral types first?